tools/api.py

```diff
@@ -0,0 +1,414 @@
+            subtrees = map(
+                lambda p: CallTree(method=p[0], order=p[1]),
+                zip(path.methods(), [0] + [edge.order for edge in path.edges()])
+            )
+            for subtree in subtrees:
+                if subtree in parent.children:
+                    parent = parent.children[parent.children.index(subtree)]
+                else:
+                    parent.children = sorted(parent.children + [subtree])
+                    parent = subtree
+        return self
+
+    def traverse(self, fn):
+        for child in self.children:
+            self._traverse_rec(fn, child)
+
+    def _traverse_rec(self, fn, root, depth=0):
+        fn(root.method, depth, not root.children)
+        for child in root.children:
+            self._traverse_rec(fn, child, depth+1)
+
+    def __lt__(self, other):
+        return self.order < other.order
+
+    def __eq__(self, other):
+        return self.method == other.method and self.order == other.order
+
+class Path:
+    def __init__(self, method, parent=None, edge=None):
+        self.method = method
+        self.parent = parent
+        self.edge = edge
+
+    def methods(self):
+        node = self
+        L = []
+        while node and node.method:
+            L.append(node.method)
+            node = node.parent
+        return list(reversed(L))
+
+    def edges(self):
+        node = self
+        L = []
+        while node and node.edge:
+            L.append(node.edge)
+            node = node.parent
+        return list(reversed(L))
+
+    def to(self, method, edge):
+        return Path(method, parent=self, edge=edge)
+
+def tree_print(method, depth, isendpoint):
+    print("{0}{1}".format('\t' * depth, method))
+
+def read_endpoints_from_api(api):
+    endpoints = read_endpoints_from_swagger_yaml(api.path, api.package)
+    for ep in endpoints:
+        ep.api = api
+    return endpoints
+
+def main(argv):
+    """Program entry point."""
+    EXIT_SUCCESS = 0
+    EXIT_FAILURE = 1
+    if len(argv) != 3:
+        sys.stderr.write("Usage: python {0} project-root callgraph-file\n".format(argv[0]))
+        return EXIT_FAILURE
+    project_root = argv[1]
+    callgraph_path = argv[2]
+    apis = [
+        API("Workbench", os.path.join(project_root, "api/src/main/resources/workbench.yaml"), "org.pmiops.workbench.api"),
+        API("Terra", os.path.join(project_root, "api/src/main/resources/fireCloud.yaml"), "org.pmiops.workbench.firecloud.api"),
+        API("Notebooks", os.path.join(project_root, "api/src/main/resources/notebooks.yaml"), "org.pmiops.workbench.notebooks.api"),
+        API("JIRA", os.path.join(project_root, "api/src/main/resources/jira.yaml"), "org.pmiops.workbench.jira.api"),
+        API("Mandrill", os.path.join(project_root, "api/src/main/resources/mandrill_api.yaml"), "org.pmiops.workbench.mandrill.api"),
+        API("Moodle", os.path.join(project_root, "api/src/main/resources/moodle.yaml"), "org.pmiops.workbench.moodle.api"),
```

Comment:
Do we need to manually add in APIs when we add them here? If we do, what happens if an API isn't included here? Is it just ignored?

This is something I could see us needing to edit, probably more than anything else in the file. Can we make it easily discoverable? (Either by moving to a variable at the top of the file, or throwing some sort of obvious comment?)

Reply:
The program will crash if a yaml file is unreadable. Your idea is good -- I'll move this list as near the top of the file as I can, and print a warning if a file is unreadable. If we find that this tool is useful for other repositories then I think it would be wise to soft-code the yaml file paths.